src/test/java/com/artipie/npm/proxy/NpmProxyITCase.java

```diff
@@ -0,0 +1,237 @@
+            "install",
+            "asdas"
+        );
+        MatcherAssert.assertThat(
+            result.getStdout(),
+            new AllOf<>(
+                Arrays.asList(
+                    new StringContains("+ asdas@1.0.0"),
+                    new StringContains("added 1 package")
+                )
+            )
+        );
+    }
+
+    @Test
+    public void packageNotFound() throws IOException, InterruptedException {
+        final Container.ExecResult result = this.npmcnter.execInContainer(
+            "npm",
+            "--registry",
+            "http://host.testcontainers.internal:9080",
+            "install",
+            "asdas"
+        );
+        MatcherAssert.assertThat(result.getExitCode(), new IsEqual<>(1));
+        MatcherAssert.assertThat(
+            result.getStderr(),
+            new StringContains(
+                "Not Found - GET http://host.testcontainers.internal:9080/asdas"
+            )
+        );
+    }
+
+    @Test
+    public void assetNotFound() throws IOException, InterruptedException {
+        final MockServerClient mock = new MockServerClient(
+            this.mockctner.getContainerIpAddress(),
+            this.mockctner.getServerPort()
+        );
+        this.successPackage(mock);
+        final Container.ExecResult result = this.npmcnter.execInContainer(
+            "npm",
+            "--registry",
+            "http://host.testcontainers.internal:9080",
+            "install",
+            "asdas"
+        );
+        MatcherAssert.assertThat(result.getExitCode(), new IsEqual<>(1));
+        MatcherAssert.assertThat(
+            result.getStderr(),
+            new StringContains(
+                "Not Found - GET http://host.testcontainers.internal:9080/asdas/-/asdas-1.0.0.tgz"
+            )
+        );
+    }
+
+    @BeforeEach
+    void setUp() {
+        final String address = this.mockctner.getContainerIpAddress();
+        final Integer port = this.mockctner.getFirstMappedPort();
+        final BaseConfiguration config = new BaseConfiguration();
+        config.setProperty("ssl", false);
+        config.setProperty("host", address);
+        config.setProperty("port", port);
+        final NpmProxySettings settings = new NpmProxySettings(config);
+        final Storage storage = new InMemoryStorage();
+        final NpmProxy npm = new NpmProxy(settings, NpmProxyITCase.VERTX, storage);
+        final NpmProxySlice slice = new NpmProxySlice(npm);
+        this.srv = new VertxSliceServer(NpmProxyITCase.VERTX, slice, NpmProxyITCase.DEF_PORT);
+        this.srv.start();
+    }
+
+    @AfterEach
+    void tearDown() {
+        this.srv.stop();
+    }
+
+    @BeforeAll
+    static void prepare() {
+        Testcontainers.exposeHostPorts(NpmProxyITCase.DEF_PORT);
+    }
+
+    @AfterAll
+    static void finish() {
+        NpmProxyITCase.VERTX.close();
+    }
+
+    private void successPackage(final MockServerClient mock) throws IOException {
+        mock.when(
```

Comment:
Why do you use this weird mock client? Can't you just any real npm sever? Like [Verdaccio](https://github.com/verdaccio/verdaccio)? It has a [docker image](https://hub.docker.com/r/verdaccio/verdaccio/) by the way.

Reply:
I have created puzzle